### PR TITLE
Match ergonomics 2024: align with RFC again

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
- `&` matches `&mut` on old editions
- Add some more tests

r? @Nadrieril

cc https://github.com/rust-lang/rust/issues/123076

@rustbot label A-edition-2024 A-patterns

### CODE REVIEW

File: tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2021.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+//@ run-pass
+//@ edition: 2021
+//@ revisions: classic structural both
+#![allow(incomplete_features)]
+#![cfg_attr(any(classic, both), feature(ref_pat_eat_one_layer_2024))]
+#![cfg_attr(any(structural, both), feature(ref_pat_eat_one_layer_2024_structural))]
+
+pub fn main() {
+    if let &Some(Some(x)) = &Some(&mut Some(0)) {
+        let _: &u32 = x;
```
Comment (traviscross) [inline tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2021.rs:10]: This test would pass even if we ended up with a `&mut u32` here, e.g.:

```rust
let x = &mut ();
let _: &() = x; //~ OK
```

Comment (Jules-Bertholet) [reply tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2021.rs:10]: It would fail borrowck as `&mut` is not `Copy`

Comment (traviscross) [reply tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2021.rs:10]: But note that:

```rust
    let x = &mut ();
    let _: &&() = &x; //~ ERROR mismatched types
```

So perhaps we could write:

```suggestion
        let _: &&u32 = &x;
```

Comment (traviscross) [reply tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2021.rs:10]: Makes sense.

**Review by traviscross — APPROVED** [review]
Looks good to me.

**Comment by Nadrieril** [comment]
To me too!

bors r plus

**Pull request merged**